File: engine/src/gadgets.c
#include "engine.h"

#include <assert.h>
#include <string.h>

static int printIndex;

void ResetPrint (void) {
    printIndex = 0;
    memset(g_PrintBuffer, 0, NMSGS * sizeof(Message));
}

static void PrintHandler (Gadget* gp, int inlet, Message msg) {
    assert(inlet == 0);

    Message arg = *(Message*) ExtraData(gp);
    if (arg != 0 && printIndex < NMSGS)
        g_PrintBuffer[printIndex++] = arg;
    if (printIndex < NMSGS)
        g_PrintBuffer[printIndex++] = msg;
}

static Gadget* MakePrintGadget (Message msg) {
    Gadget* gp = NewGadget(sizeof(Message), PrintHandler);
    *(Message*) ExtraData(gp) = msg;
    return gp;
}

static Gadget* MakePassGadget (Message msg) {
    (void) msg;
    return NewGadget(0, Emit);
}

static Gadget* MakeInletGadget (Message msg) {
    (void) msg;
    return NewGadget(sizeof(Wire), 0);
}

static void OutletHandler (Gadget* gp, int inlet, Message msg) {
    assert(inlet == 0);

    // scan the sibling gadgets to find the matching outlet
    Gadget **gpp = ExtraData(gp->parent);
    int outlet = 0;
    while (*gpp != gp)
        if ((*gpp++)->handler == OutletHandler)
            ++outlet;

    Emit(gp->parent, outlet, msg);
}

static Gadget* MakeOutletGadget (Message msg) {
    (void) msg;
    return NewGadget(sizeof(Wire), OutletHandler);
}

static void SwapHandler (Gadget* gp, int inlet, Message msg) {
    switch (inlet) {
        case 0:
            Emit(gp, 1, msg);
            Emit(gp, 0, *(Message*) ExtraData(gp));
            break;
        case 1:
            *(Message*) ExtraData(gp) = msg;
            break;
        default:
            assert(0);
    }
}

static Gadget* MakeSwapGadget (Message msg) {
    Gadget* gp = NewGadget(sizeof(Message), SwapHandler);
    *(Message*) ExtraData(gp) = msg;
    return gp;
}

static void ChangeHandler (Gadget* gp, int inlet, Message msg) {
    assert(inlet == 0);

    Message arg = *(Message*) ExtraData(gp);
    if (msg != arg) {
        *(Message*) ExtraData(gp) = msg;
        Emit(gp, 0, msg);
    }
}

static Gadget* MakeChangeGadget (Message msg) {
    (void) msg;
    Gadget* gp = NewGadget(sizeof(Message), ChangeHandler);
    *(Message*) ExtraData(gp) = -1;
    return gp;
}

static void MosesHandler (Gadget* gp, int inlet, Message msg) {
    assert(inlet == 0);

    Message arg = *(Message*) ExtraData(gp);
    Emit(gp, msg >= arg, msg);
}

static Gadget* MakeMosesGadget (Message msg) {
    Gadget* gp = NewGadget(sizeof(Message), MosesHandler);
    *(Message*) ExtraData(gp) = msg;
    return gp;
}

struct Lookup_t g_Gadgets[] = {
    { "print",  MakePrintGadget  },
    { "pass",   MakePassGadget   },
    { "inlet",  MakeInletGadget  },
    { "outlet", MakeOutletGadget },
    { "swap",   MakeSwapGadget   },
    { "change", MakeChangeGadget },
    { "moses",  MakeMosesGadget  },
    { 0, 0 }
};

File: flow/gadgets.fs
\ <<<engine>>>
\ cr compiletoflash

: m. ( m -- ) . ;
: m! ( m a -- ) ! ;

: :print ( arg -- g )
  [: ( msg in -- ) drop  extra @ ?dup if m. then m. ;]
  1 cells new-gadget ( arg g )
  tuck g-extra m! ;

: :pass ( arg -- g )
  drop
  [: ( msg in -- ) g-emit ;]
  0 new-gadget ;

: :inlet ( arg -- g )
  drop  0 0 new-gadget ;

: outlet-h ( msg in -- )
  parent g-extra
  begin ( msg n gpp )
    dup @ cg @ <> while
    swap 1+ swap
    cell+
  repeat drop
  cg @ >r parent cg ! g-emit r> cg ! ;

: :outlet ( arg -- g )
  drop  ['] outlet-h 0 new-gadget ;

: swap-h ( msg in -- )
  0= if
    1 g-emit
    extra @ 0 g-emit
  else
    extra m!
  then ;

: :swap ( arg -- g )
  ['] swap-h 1 cells new-gadget ( arg g )
  tuck g-extra m! ;

: change-h ( msg in -- )
  drop dup extra @ <> if
    dup extra !
    dup 0 g-emit
  then drop ;

: :change ( arg -- g )
  drop  ['] change-h 1 cells new-gadget ( arg g )
  -1 i>m extra m! ;

: moses-h ( msg in -- )
  drop dup extra @ < if 0 else 1 then g-emit ;

: :moses ( arg -- g )
  ['] moses-h 1 cells new-gadget ( arg g )
  tuck g-extra m! ;
